### ISSUE TITLE
Implement a Query that matches a predicate over fast fields.

### ISSUE THREAD

**Issue by fulmicoton** [quickwit-oss/tantivy#396]
**Is your feature request related to a problem? Please describe.**

Would like to be able to filter over a fastfield. 

**Describe the solution you'd like**

For the first implementation, we won't precompute/cache any bitset, and we 
do not need to worry about optimizing its association with deletes.
A straightforward `Query`, `Weight`, `Scorer` is ok.

**[Optional] describe alternatives you've considered**

All of the following implementation are out of scope of this ticket but are relevant 
features.

- Range query make it possible to filter over range of terms, but they are currently quite inefficient as we do not do the radix decomposition of the integer.
- We do not have the two-pass query mechanics yet, that would have helped.
- Filter should eventually be cachable as a bitset and get merged with the delete bitset.

**Comment by ddorian** [comment]
In what scenario would this be better compared to indexing the field ? In cases where you query/filter rarely on that field so it costs more to search on fast-field instead of write+search on indexed-field ?

**Comment by fulmicoton** [comment]
Well for this specific case, with the current version of tantivy, it is almost always faster...
Currently, range queries (1970 <= year < 2018) are done by expanding all terms between the two boundaries. This can be a lot of terms...

There are different ways to solve this problem. Doing a n-bit decomposition of the values for instance.
It prevents the explosition of the number of terms to search.

Even then... filtering over the field will always can be faster IF the ratio of positives is high.

In fact, more fundamentally, even if A is a TermQuery and regardless of what B is; the fastest way to compute a DocSet for  A intersection B when B is filtering away very few element of A, is to have B as bitset.

The limit ratio when it becomes a better idea is somewher between 1/6 and 1/3 I think.

**Comment by kkonevets** [comment]
What is a work around for now?

For example, I want to filter docs which have "company_id == 33" and take a top 10 with highest score, how can I do it for now?

Should I implement custom collector for this? For example, int top facet collector where I filter docs matching a predicate on facet filed. I guess I should implement `collect()` method where I skip adding doc to a binary heap if its fast field does't match a predicate (e.g. "company_id == 33")?

**Comment by kkonevets** [comment]
how can I make a select query?
I want to get top 10 docs ordered by score where "test_i64==47". Should test_i64 be a fast field?

Should I use RangeQuery and TermQuery combined with BooleanQuery? As far as documentation for RangeQuery sates "Matched document will all get a constant Score of one", so how can I then get top 10? Is RangeQuery weight multiplied by TermQuery weight? If yes, it makes sense. Is there an example of such a query?

Or should I implement custom collector for that in which I will add docs only with predicate "test_i64==47" ?

**Comment by fulmicoton** [comment]
> I want to get top 10 docs ordered by score 

I am missing something... What is your score here? 
Tantivy can use a fast field as a score, or build a BM25-similarity score if you have a query...

If your query is just test_i64==47, BM25 score will not be super relevant I'm afraid.

If you have a normal text query like "bluetooth headphone ", and you want to furthermore filter it by   `test_i64==47`, then yes, you can configure `test_i64` as indexed in your schema and combine the two in a BooleanQuery. 

If your filter is a range, or something more complicated, it can absolutely make sense to implement a custom collector wrapping another collector, and filtering documents that should be pushed to it.

**Comment by kkonevets** [comment]
Thank you for the answer. Could you please explain a little more?

I use a default score. Is it BM25 or tf-idf? How can I set either one manually? I could't find a single word about BM25 or tf-idf on the package documentation (docs.rs web site). basic_search.rs example states that by default it uses modified tf-idf score, how can I set BM25? Or by "modified tf-idf score" you mean BM25?

Yes I have a text query and a select condition, e.g. find 'bluetooth headphone' where `company_id==47` and give me top 10 most relevant.

If I just want to filter by `company_id==47` should I use a range query such as `company_id: [47 TO 47]`? Or a query like `company_id: 47` will be just fine?

Why should I implement a custom collector for a range query? QueryParser allows us to use range queries with a TopCollector. As far as I know query gets documents and only after that collector collects those, this means that it will be faster to delegate predicate condition (`company_id==47`) to the query and not to the collector.

**Comment by fulmicoton** [comment]
> I use a default score. Is it BM25 or tfidf?

BM25

> If I just want to filter by company_id==47 should I use range query such as company_id: [47 TO 47]? I > couldn't find == query syntax.

It's just a `TermQuery`. You can have TermQuery over i64 field.

```rust
let term = Term::from_field_i64(test_i64, 47i64);
let term_filtering_query = TermQuery::new(my_term, IndexRecordOption::BASIC);
```

The query parser should also understand `company_id: 47` just fine.

> Why should I implement a custom collector for a range query? QueryParser allows us to use range queries with a TopCollector. As far as I know query gets documents and only after that collector collects those, this means that it will be faster to delegate predicate condition (company_id==47) to the query and not to the collector.

If you filter for a specific value it is useless yes. Sometimes if you filter for a range it will be faster to use a custom collector.

**Comment by PSeitz** [comment]
That has been implemented in the meantime via `FastFieldRangeQuery`
